### ISSUE TITLE
error[E0505]: cannot move out of `socket` because it is borrowed   --> src/macros.rs:21:29

### ISSUE THREAD

**Issue by tristanpoland** [Stars-Beyond/Horizon-Community-Edition#80]
### Description

Compiler asks me to clone my socket variable when using the defineevent macro, but it seems inefficient to do this for *every* event I want to define:

```bash
   |
34 |         define_event!(socket.clone(), "command", handle_help(socket.clone(), ""));
   |                             ++++++++
   ```

Sample code that causes the issue:

```rs
//////////////////////////////////////
// Lets import some tools we need.  //
//////////////////////////////////////

use socketioxide::extract::{SocketRef, Data};
use serde::Deserialize;
use tracing::info;
use crate::define_event;

////////////////////////////////////////////////
// Next we will define some structs our code  //
// will use to store and organize data.       //
////////////////////////////////////////////////

#[derive(Deserialize)]
struct ChatMessage {
    sender: String,
    recipient: Option<String>,
    message: String,
}

/////////////////////////////////////////////////
//  Next we will define what happens when our  //
//  subsystem starts, this usually happens     //
//  when a new player connects, so we will     //
//  define our event listeners here.           //
/////////////////////////////////////////////////

pub fn init(socket: SocketRef) {
    info!("Starting chat subsystem...");
    
        define_event!(socket, "whisper", handle_whisper(socket.clone(), "", "", ""));
        define_event!(socket, "broadcast", handle_broadcast(socket.clone(), "", ""));
        define_event!(socket, "command", handle_help(socket.clone(), ""));
    
}
    
    ////////////////////////////////////////////////////
    //  Finally we define the functions that will be  //
    //  called by our event listeners when they are   //
    //  triggered.                                    //
    ////////////////////////////////////////////////////
    
    fn handle_whisper(socket: SocketRef, sender: &str, recipient: &str, message: &str) {
        info!("Whisper from {} to {}: {}", sender, recipient, message);
        // Example of emitting a message to a specific recipient
        socket.emit("whisper", (recipient, message)).ok();
    }
    
    fn handle_help(socket: SocketRef, sender: &str) {
        info!("Help requested by {}", sender);
        // Example of emitting a help message to the sender
        socket.emit("help", "Here's some help!").ok();
    }
    
    fn handle_broadcast(socket: SocketRef, sender: &str, message: &str) {
        info!("Broadcast from {}: {}", sender, message);
        // Example of broadcasting a message to all connected clients
        socket.broadcast().emit("broadcast", message).ok();
    }
    ```

### Reproduction steps

```bash
1. Checkout this branch: [origin/defineevent-macro-bugfix](https://github.com/Stars-Beyond/Horizon-Community-Edition/tree/defineevent-macro-bugfix)
2. Open project
3. do ``cargo run``
```

### Screenshots

```bash
https://gist.github.com/user-attachments/assets/648edc2d-3838-4a87-9bac-8f398179496b
https://gist.github.com/user-attachments/assets/22460d83-f1cf-44a7-bf2c-0c6ae6b2c073
https://gist.github.com/user-attachments/assets/3553360d-923e-4529-840b-b6c2ff47dc7a
```

### Logs

```bash
vscode ➜ /workspaces/Horizon-Community-Edition (defineevent-macro-bugfix) $ cargo run
warning: use of deprecated constant `bindgen::CargoCallbacks`: Use `CargoCallbacks::new()` instead. Please, check the documentation for further information.
  --> src/crates/PebbleVault/build.rs:31:44
   |
31 |         .parse_callbacks(Box::new(bindgen::CargoCallbacks))
   |                                            ^^^^^^^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

warning: `PebbleVault` (build script) generated 1 warning
   Compiling PebbleVault v0.1.0 (/workspaces/Horizon-Community-Edition/src/crates/PebbleVault)
warning: unused import: `self`
 --> src/crates/TerraForge/src/delaunay_triangulation.rs:3:15
  |
3 | use std::io::{self, Write};
  |               ^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: crate `TerraForge` should have a snake case name
  |
  = help: convert the identifier to snake case: `terra_forge`
  = note: `#[warn(non_snake_case)]` on by default

   Compiling Horizon v0.1.0 (/workspaces/Horizon-Community-Edition)
warning: `TerraForge` (lib) generated 2 warnings (run `cargo fix --lib -p TerraForge` to apply 1 suggestion)
warning: unused import: `MySQLGeo::Database`
  --> src/crates/PebbleVault/src/lib.rs:12:5
   |
12 | use MySQLGeo::Database;
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused variable: `data_str`
  --> src/crates/PebbleVault/src/MySQLGeo.rs:79:13
   |
79 |         let data_str = serde_json::to_string(&point.data)
   |             ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_data_str`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `Result`
   --> src/crates/PebbleVault/src/MySQLGeo.rs:105:13
    |
105 |         let Result: Result<(), std::io::Error> = fs::write(&filePath, contents);
    |             ^^^^^^ help: if this is intentional, prefix it with an underscore: `_Result`

warning: unused variable: `data`
  --> src/crates/PebbleVault/src/lib.rs:41:38
   |
41 |     pub fn collect(&self, key: &str, data: &str) {
   |                                      ^^^^ help: if this is intentional, prefix it with an underscore: `_data`

warning: unused variable: `data`
  --> src/crates/PebbleVault/src/lib.rs:67:48
   |
67 |     pub fn pebbledump(&self, table_name: &str, data: Vec<&str>) {
   |                                                ^^^^ help: if this is intentional, prefix it with an underscore: `_data`

warning: unused variable: `new_data`
  --> src/crates/PebbleVault/src/lib.rs:72:42
   |
72 |     pub fn pebbleshift(&self, key: &str, new_data: &str) {
   |                                          ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_new_data`

warning: unused variable: `query_conditions`
  --> src/crates/PebbleVault/src/lib.rs:77:48
   |
77 |     pub fn pebblesift(&self, table_name: &str, query_conditions: &str) -> Vec<String> {
   |                                                ^^^^^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_query_conditions`

warning: unused variable: `partial_data`
  --> src/crates/PebbleVault/src/lib.rs:83:42
   |
83 |     pub fn pebblepatch(&self, key: &str, partial_data: &str) {
   |                                          ^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_partial_data`

warning: unused variable: `data`
   --> src/crates/PebbleVault/src/lib.rs:115:38
    |
115 |     pub fn collect(&self, key: &str, data: &str) {
    |                                      ^^^^ help: if this is intentional, prefix it with an underscore: `_data`

warning: fields `mem_db_handle` and `sql_db_handle` are never read
  --> src/crates/PebbleVault/src/lib.rs:22:5
   |
21 | pub struct Vault {
   |            ----- fields in this struct
22 |     mem_db_handle: *mut c_void,
   |     ^^^^^^^^^^^^^
23 |     sql_db_handle: MySQLGeo::Database,
   |     ^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: associated function `new` is never used
  --> src/crates/PebbleVault/src/MySQLGeo.rs:21:12
   |
20 | impl Point {
   | ---------- associated function in this implementation
21 |     pub fn new(id: Option<Uuid>, x: f64, y: f64, z: f64, data: Value) -> Self {
   |            ^^^

warning: methods `add_point` and `get_points_within_radius` are never used
   --> src/crates/PebbleVault/src/MySQLGeo.rs:78:12
    |
26  | impl Database {
    | ------------- methods in this implementation
...
78  |     pub fn add_point(&self, point: &Point) -> SqlResult<()> {
    |            ^^^^^^^^^
...
122 |     pub fn get_points_within_radius(&self, x1: f64, y1: f64, z1: f64, radius: f64) -> SqlResult<Vec<Point>> {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused `Result` that must be used
  --> src/crates/PebbleVault/src/lib.rs:33:9
   |
33 |         MySQLGeo::Database::create_table(&sql_db_handle);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this `Result` may be an `Err` variant, which should be handled
   = note: `#[warn(unused_must_use)]` on by default
help: use `let _ = ...` to ignore the resulting value
   |
33 |         let _ = MySQLGeo::Database::create_table(&sql_db_handle);
   |         +++++++

warning: unused `Result` that must be used
   --> src/crates/PebbleVault/src/lib.rs:160:5
    |
160 |     MySQLGeo::Database::create_table(&sql_db_handle);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this `Result` may be an `Err` variant, which should be handled
help: use `let _ = ...` to ignore the resulting value
    |
160 |     let _ = MySQLGeo::Database::create_table(&sql_db_handle);
    |     +++++++

warning: `PebbleVault` (lib) generated 14 warnings (run `cargo fix --lib -p PebbleVault` to apply 1 suggestion)
warning: unused import: `http::status`
  --> src/main.rs:24:5
   |
24 | use http::status;
   |     ^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `future::ok`
  --> src/main.rs:30:11
   |
30 | use viz::{future::ok, handler::ServiceHandler, serve, Response, Result, Router, Request, Body};
   |           ^^^^^^^^^^

warning: unused import: `Data`
 --> src/subsystems/core/chat.rs:5:40
  |
5 | use socketioxide::extract::{SocketRef, Data};
  |                                        ^^^^

error[E0382]: borrow of moved value: `socket`
   --> src/macros.rs:21:14
    |
21  |             &$app.on($path, move || { $handler });
    |              ^^^^^^^^^^^^^^^-------^^^^^^^^^^^^^^
    |              |              |
    |              |              value moved into closure here
    |              value borrowed here after move
    |
   ::: src/subsystems/core/chat.rs:29:13
    |
29  | pub fn init(socket: SocketRef) {
    |             ------ move occurs because `socket` has type `SocketRef`, which does not implement the `Copy` trait
...
32  |         define_event!(socket, "whisper", handle_whisper(socket.clone(), "", "", ""));
    |                                                         ------ variable moved due to use in closure
33  |         define_event!(socket, "broadcast", handle_broadcast(socket.clone(), "", ""));
    |         ---------------------------------------------------------------------------- in this macro invocation
    |
    = note: borrow occurs due to deref coercion to `socketioxide::socket::Socket`
note: deref defined here
   --> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/socketioxide-0.13.1/src/handler/extract.rs:217:5
    |
217 |     type Target = Socket<A>;
    |     ^^^^^^^^^^^
    = note: this error originates in the macro `define_event` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0382]: borrow of moved value: `socket`
   --> src/macros.rs:21:14
    |
21  |             &$app.on($path, move || { $handler });
    |              ^^^^^^^^^^^^^^^-------^^^^^^^^^^^^^^
    |              |              |
    |              |              value moved into closure here
    |              value borrowed here after move
    |
   ::: src/subsystems/core/chat.rs:29:13
    |
29  | pub fn init(socket: SocketRef) {
    |             ------ move occurs because `socket` has type `SocketRef`, which does not implement the `Copy` trait
...
33  |         define_event!(socket, "broadcast", handle_broadcast(socket.clone(), "", ""));
    |                                                             ------ variable moved due to use in closure
34  |         define_event!(socket, "command", handle_help(socket.clone(), ""));
    |         ----------------------------------------------------------------- in this macro invocation
    |
    = note: borrow occurs due to deref coercion to `socketioxide::socket::Socket`
note: deref defined here
   --> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/socketioxide-0.13.1/src/handler/extract.rs:217:5
    |
217 |     type Target = Socket<A>;
    |     ^^^^^^^^^^^
    = note: this error originates in the macro `define_event` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0505]: cannot move out of `socket` because it is borrowed
  --> src/macros.rs:21:29
   |
21 |             &$app.on($path, move || { $handler });
   |                   --        ^^^^^^^ move out of `socket` occurs here
   |                   |
   |                   borrow later used by call
   |
  ::: src/subsystems/core/chat.rs:29:13
   |
29 | pub fn init(socket: SocketRef) {
   |             ------ binding `socket` declared here
...
32 |         define_event!(socket, "whisper", handle_whisper(socket.clone(), "", "", ""));
   |         ----------------------------------------------------------------------------
   |         |             |                                 |
   |         |             |                                 move occurs due to use in closure
   |         |             borrow of `socket` occurs here
   |         in this macro invocation
   |
   = note: this error originates in the macro `define_event` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value if the performance cost is acceptable
  --> src/subsystems/core/chat.rs:32:29
   |
32 |         define_event!(socket.clone(), "whisper", handle_whisper(socket.clone(), "", "", ""));
   |                             ++++++++

error[E0505]: cannot move out of `socket` because it is borrowed
  --> src/macros.rs:21:29
   |
21 |             &$app.on($path, move || { $handler });
   |                   --        ^^^^^^^ move out of `socket` occurs here
   |                   |
   |                   borrow later used by call
   |
  ::: src/subsystems/core/chat.rs:29:13
   |
29 | pub fn init(socket: SocketRef) {
   |             ------ binding `socket` declared here
...
33 |         define_event!(socket, "broadcast", handle_broadcast(socket.clone(), "", ""));
   |         ----------------------------------------------------------------------------
   |         |             |                                     |
   |         |             |                                     move occurs due to use in closure
   |         |             borrow of `socket` occurs here
   |         in this macro invocation
   |
   = note: this error originates in the macro `define_event` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value if the performance cost is acceptable
  --> src/subsystems/core/chat.rs:33:29
   |
33 |         define_event!(socket.clone(), "broadcast", handle_broadcast(socket.clone(), "", ""));
   |                             ++++++++

error[E0505]: cannot move out of `socket` because it is borrowed
  --> src/macros.rs:21:29
   |
21 |             &$app.on($path, move || { $handler });
   |                   --        ^^^^^^^ move out of `socket` occurs here
   |                   |
   |                   borrow later used by call
   |
  ::: src/subsystems/core/chat.rs:29:13
   |
29 | pub fn init(socket: SocketRef) {
   |             ------ binding `socket` declared here
...
34 |         define_event!(socket, "command", handle_help(socket.clone(), ""));
   |         -----------------------------------------------------------------
   |         |             |                              |
   |         |             |                              move occurs due to use in closure
   |         |             borrow of `socket` occurs here
   |         in this macro invocation
   |
   = note: this error originates in the macro `define_event` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value if the performance cost is acceptable
  --> src/subsystems/core/chat.rs:34:29
   |
34 |         define_event!(socket.clone(), "command", handle_help(socket.clone(), ""));
   |                             ++++++++

warning: unused variable: `data`
  --> src/main.rs:55:39
   |
55 | fn on_connect(socket: SocketRef, Data(data): Data<Value>, players: Arc<Mutex<Vec<Player>>>) {
   |                                       ^^^^ help: if this is intentional, prefix it with an underscore: `_data`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `db`
   --> src/main.rs:225:9
    |
225 |     let db = PebbleVault::create_db();
    |         ^^ help: if this is intentional, prefix it with an underscore: `_db`

Some errors have detailed explanations: E0382, E0505.
For more information about an error, try `rustc --explain E0382`.
warning: `Horizon` (bin "horizon") generated 5 warnings
error: could not compile `Horizon` (bin "horizon") due to 5 previous errors; 5 warnings emitted
```

### OS

Windows, Linux, Mac, Steam Deck

**Comment by WilliamAnimate** [comment]
yowch. the ownership system is right.
in rust, you can only use a value once (moving ownership) or borrowing the value for others (which allows you to use it infinite times, until the owner changes)*

since we're in a multi threaded environment, we may need to share ownership of the `SocketRef` type, so we may need to import the `lazy_static` or `once_cell` crate and then wrap it around in an `std::sync::arc::Arc`. doing that, we can "clone" the value to use it. cloning a value wrapped within an Arc is very cheap, because it doesn't memcpy the entire value (?) but increments the reference count

this may need reworking of how some things work.

*something like that. I'm out of the country so my brain is out of order

**Comment by WilliamAnimate** [comment]
disregard everything I've said. an easier solution would be to make the functions or methods accept a borrowed SocketRef

like:

```rust
fn on_connect(socket: &SocketRef) {}
```

**Comment by WilliamAnimate** [comment]
okay, I just reproduced the issue on my end. give me some time to figure out how to make rustc happy

just give me a bit; I've only got a mobile phone to work with

**Comment by tristanpoland** [comment]
@WilliamAnimate 

Absolute mad lad!!!

**Comment by WilliamAnimate** [comment]
upon further inspection, I see what is happening

I assumed the macro itself was broken (because I did nothing on commit 211b5a2), but it turns out we are trying to pass the SocketRef into a function, which causes problems

**Comment by WilliamAnimate** [comment]
hit close by accident my bad

**Comment by tristanpoland** [comment]
> hit close by accident my bad

That was me actually 💀 oops

**Comment by WilliamAnimate** [comment]
> > hit close by accident my bad
> 
> That was me actually 💀 oops

lmfao my view shifted when it closed which made me think I scrolled (and touched the screen)

**Comment by WilliamAnimate** [comment]
> upon further inspection, I see what is happening
> 
> I assumed the macro itself was broken (because I did nothing on commit [211b5a2](https://github.com/Stars-Beyond/Horizon-Community-Edition/commit/211b5a20be17ba8e650925cbc7666879afeec1ac)), but it turns out we are trying to pass the SocketRef into a function, which causes problems

![Screenshot_20240713_204508_Termux](https://github.com/user-attachments/assets/6baf0edd-6a27-4bbd-84ad-8677950a0ad3)

we *could* clone here and that might fix the issue, but that defeats the point, so it may be worthwhile to use Arc here?

going to try to find ways to work around this before I try refactoring

**Comment by tristanpoland** [comment]
> > upon further inspection, I see what is happening
> > I assumed the macro itself was broken (because I did nothing on commit [211b5a2](https://github.com/Stars-Beyond/Horizon-Community-Edition/commit/211b5a20be17ba8e650925cbc7666879afeec1ac)), but it turns out we are trying to pass the SocketRef into a function, which causes problems
> 
> ![Screenshot_20240713_204508_Termux](https://private-user-images.githubusercontent.com/75321465/348472484-6baf0edd-6a27-4bbd-84ad-8677950a0ad3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA4NzUxODEsIm5iZiI6MTcyMDg3NDg4MSwicGF0aCI6Ii83NTMyMTQ2NS8zNDg0NzI0ODQtNmJhZjBlZGQtNmEyNy00YmJkLTg0YWQtODY3Nzk1MGEwYWQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEzVDEyNDgwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlODJiZjY2ZTM1ZDY0NjAzZGNiNTBmOWY4ZDM1MjEyYjQ2NTVlNWRkZDExOGViMjU0ODc3MWI2NGZkMzZjZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oIQQc6OhifWwWeoo2b1b7cpF7K0NguuDIbzgDXGdJyg)
> 
> we _could_ clone here and that might fix the issue, but that defeats the point, so it may be worthwhile to use Arc here?
> 
> going to try to find ways to work around this before I try refactoring

For contenxt we *could* have thousands or tens of thousands of events in a very large game

**Comment by tristanpoland** [comment]
Implemented a fix for this in #88 I hope to find a more memory efficient one later but this should technically be perfectly fine
